File: OneStepAheadResiduals/OSA.cpp
// Code from Thygesen et al., 2017 is used.
#include <TMB.hpp>
template <class Type>
Type objective_function<Type>::operator()()
{
  DATA_VECTOR(y);                  //Observations
  DATA_VECTOR_INDICATOR(keep, y);  //Indicator variable for one-step-ahead predictions
  DATA_SCALAR(huge);    //Used to integreat out X_1
  PARAMETER_VECTOR(x);  //The latent AR1 random variable
  PARAMETER(mu);        //The drift
  PARAMETER(logsigma);  //Log sd of epsilon
  PARAMETER(logs);      //Log sd of psi
  
  // Initial condition
  Type nll = -dnorm(x(0), Type(0), huge, true);
  
  // Increments
  for (int i = 1; i < x.size(); ++i)
    nll -= dnorm(x(i), x(i - 1) + mu, exp(logsigma), true);
  
  // Observations
  for (int i = 0; i < y.size(); ++i)
    nll -= keep(i) * dnorm(y(i), x(i), exp(logs), true);
  
  return nll;
}

File: OneStepAheadResiduals/OSA.R
# Code from Thygesen et al., 2017 is used.
library(TMB)
compile("OSA.cpp")
dyn.load(dynlib("OSA"))

set.seed(123)
## Simulate data with these parameters
mu = 0.75
sigma = 1
s = 1
huge = 1e3
nT = 100
X = c(0,cumsum(rnorm(nT-1,mean=mu,sd=sigma)))
Y = X + rnorm(nT,sd=s)

data = list(y=Y,huge=huge)
parameters = list(x=X,mu=0,logsigma=log(sigma),logs=log(s))
obj = MakeADFun(data,parameters,random=c("x"),DLL="OSA")
opt = nlminb(obj$par, obj$fn,obj$gr)


### One-step predictions
predict  <- oneStepPredict(obj,observation.name="y",method="fullGaussian",data.term.indicator = "keep")
plot(predict$residual,xlab="Time", main = "OSA residuals", ylab = "Residual")
qqnorm(predict$residual)
qqline(predict$residual, col = "steelblue", lwd = 2)
